agent/resources_monitor.go

```diff
@@ -0,0 +1,135 @@
+package agent
+
+import (
+	"context"
+	"time"
+
+	"golang.org/x/xerrors"
+
+	"cdr.dev/slog"
+	"github.com/coder/coder/v2/agent/proto"
+	"github.com/coder/coder/v2/cli/clistat"
+	"github.com/coder/quartz"
+)
+
+func (a *agent) pushResourcesMonitoring(ctx context.Context, aAPI proto.DRPCAgentClient24) error {
+	logger := a.logger.Named("resources_monitor")
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	clk := quartz.NewReal()
+
+	return PushResourcesMonitoringWithConfig(ctx, logger, clk, aAPI.GetResourcesMonitoringConfiguration, aAPI.PushResourcesMonitoringUsage)
+}
+
+type (
+	ResourcesMonitorConfigurationFetcher func(ctx context.Context, params *proto.GetResourcesMonitoringConfigurationRequest) (*proto.GetResourcesMonitoringConfigurationResponse, error)
+	ResourcesMonitorDatapointsPusher     func(ctx context.Context, params *proto.PushResourcesMonitoringUsageRequest) (*proto.PushResourcesMonitoringUsageResponse, error)
+)
+
+func PushResourcesMonitoringWithConfig(ctx context.Context,
+	logger slog.Logger,
+	clk quartz.Clock,
+	configFetcher ResourcesMonitorConfigurationFetcher,
+	datapointsPusher ResourcesMonitorDatapointsPusher,
+) error {
+	config, err := configFetcher(ctx, &proto.GetResourcesMonitoringConfigurationRequest{})
```

Comment:
self review : There's three situations on the which we do not want to run the go-routine : 
- If we can not fetch the configuration from Coderd.
- If the resources_monitoring is not enabled.
- If we can not instantiate the clistat client (which is the one used to fetch resources usage.)